File: 0198-house-robber/0198-house-robber.js
/**
 * @param {number[]} nums
 * @return {number}
 */
const rob = nums => {
    if (nums.length === 1) return nums[0];
    let max = new Array(nums.length);
    max[0] = nums[0];
    max[1] = Math.max(nums[0], nums[1]);
    for (let i = 2; i < nums.length; i++) {
        max[i] = Math.max(max[i - 1], nums[i] + max[i - 2]);
    }
    return max[max.length - 1];
};

File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        self.memo = {}
        return self.robFrom(0, nums)
    def robFrom(self, i, nums):
        if i >= len(nums):
            return 0
        if i in self.memo:
            return self.memo[i]
        ans = max(nums[i] + self.robFrom(i + 2, nums), self.robFrom(i + 1, nums))
        self.memo[i] = ans
        return self.memo[i]